### ISSUE TITLE
Slice shape inference panics on i64::MIN sentinel (overflow in normalize_index's idx+n)

### ISSUE THREAD

**Issue by krishna3554** [tracel-ai/burn-onnx#476]
## Summary

`calculate_dim_slice_output_len()` in `crates/onnx-ir/src/node/slice.rs` normalizes negative indices with `idx + n` using plain addition:

```rust
fn normalize_index(idx: i64, lo: i64, hi: i64, n: i64) -> i64 {
    let abs = if idx < 0 { idx + n } else { idx };
    abs.clamp(lo, hi)
}
```

The ONNX Slice spec defines **`i64::MIN` as an explicit sentinel** (for reverse slicing, `end = INT64_MIN` means "before the first element"). When such a sentinel reaches this helper, `idx + n` overflows `i64`: debug builds **panic** ("attempt to add with overflow") during model import/shape inference, while release builds silently wrap and only then clamp. The function's own doc comment advertises exactly this sentinel handling ("for `step < 0`, `i64::MIN + dim_len` clamps to `-1`"), but the arithmetic panics before the clamp can run in debug configurations.

Static-analysis finding against current `master`; verified by code reading only.

## Location

- File: `crates/onnx-ir/src/node/slice.rs`
- Functions:
  - `normalize_index()` nested inside `calculate_dim_slice_output_len()` (~lines 88-91),
  - called from `calculate_dim_slice_output_len()` (~lines 76-107), which runs at node-processing time from the Slice processor (~lines 204 and ~247).

## Problem

For a reverse slice with `ends[0] == i64::MIN` and any tensor rank:

- `step < 0` branch → `normalize_index(i64::MIN, -1, dim_len - 1, dim_len)`
- `idx < 0` → evaluates `i64::MIN + dim_len`

That add is undefined-range arithmetic in Rust's terms: debug profile aborts with an overflow panic; release profile wraps to a large negative that then clamps to `-1`. The same applies to a (spec-legal but unusual) `starts` value of `i64::MIN` on the positive-step path.

Since this code executes at import/shape-inference time inside burn-onnx, importing a perfectly valid ONNX model - one that follows the documented `INT64_MIN` convention used by PyTorch/TensorFlow exporters for reverse slices - terminates with a panic in debug builds instead of computing the output shape.

## Trigger / Reproduction

Based on static analysis; no runtime run performed:

```text
Slice node attributes/inputs:
  starts = [dim_len - 1]        (e.g. [4])
  ends   = [-9223372036854775808]
  steps  = [-1]
```

Importing this model with burn-onnx hits `normalize_index` with `idx = i64::MIN`.

## Expected Behavior

Sentinel values must be handled with saturating/checked arithmetic so the documented clamping occurs identically in all build profiles:

```rust
let abs = if idx < 0 { idx.saturating_add(n) } else { idx };
```

(or compare before adding).

## Actual Behavior

Debug builds panic with an integer-overflow abort; release builds work only incidentally via wrap+clamp, making behavior configuration-dependent.

## Impact

Any user importing a real-world model whose exporter emitted the standard `INT64_MIN` reverse-slice sentinel gets a hard panic during code generation when building with `debug_assertions` (common for local development of Burn projects). Even where it "works", relying on wrap-around is fragile.

## Suggested Direction

Switch the single addition to `saturating_add` (and audit sibling helpers in this file for the same pattern, e.g. anywhere `starts`/`ends`/`axes` are offset by rank or dim length). A unit test feeding `(i64::MIN, x, -1, n)` through `calculate_dim_slice_output_len` would pin the fix; the file header already documents the sentinel semantics to test against.

## Evidence

- Arithmetic quoted above; Rust overflow semantics (panic under `debug_assertions`, two's-complement wrap otherwise).
- ONNX Slice documentation: negative `end` values below `-rank` and the `INT64_MIN` convention represent "clamp to boundary"/"before first element".

**Comment by antimora** [comment]
Thanks for the report. The specific overflow described here doesn't happen, but chasing it down did surface a real bug in the same code path, so this is worth keeping open.

## The reported overflow is not reachable

`idx + n` in `normalize_index` is `i64::MIN + dim_len` where `dim_len >= 0`. Adding a non-negative number to `i64::MIN` moves *toward* zero, so it can't overflow:

```
i64::MIN + 8 == -9223372036854775800   // well inside i64
```

Overflow in `a + b` needs both operands to share a sign. That's a real hazard for `i64::MAX + positive`, but not for `i64::MIN + non-negative`. The doc comment ("`i64::MIN + dim_len` clamps to `-1`") is accurate, and the clamp does run.

Verified rather than reasoned: the whole `onnx-ir` slice suite passes under `RUSTFLAGS="-C overflow-checks=on"`, including `test_tensor_static_shape_end_i64_min_slices_to_start`, which already feeds exactly `(start=7, end=i64::MIN, step=-1, dim=8)` through `calculate_dim_slice_output_len`.

## Your reproduction model does fail, for a different reason

Building the model you describe (`starts=[4]`, `ends=[INT64_MIN]`, `steps=[-1]`) and running `onnx2burn` on it imports cleanly and infers the right output shape, then emits:

```rust
let slice1_out1 = x.slice(s![4.. - 9223372036854775808; - 1, ..]);
//  error: literal out of range for `i32`
```

Codegen special-cased the `i64::MAX` sentinel but not `i64::MIN`, so the sentinel reached the generated source verbatim and the generated crate didn't compile.

## Root cause underneath that

ONNX and Burn use different conventions for a negative step, and burn-onnx was passing `starts`/`ends`/`steps` straight through:

| | `start=0, end=5, step=-1` over a 5-element axis |
|---|---|
| Burn `s![0..5;-1]` | `[4, 3, 2, 1, 0]` (forward range, step sign reverses traversal) |
| ONNX | `[]` (start at 0, walk backwards, stop before 5) |

So every negative-step slice was wrong, not just the sentinel ones. `onnx-ir`'s shape inference had it right all along; codegen disagreed with it. The existing `slice_with_steps` test encoded the wrong behavior: it asserted `[5, 3, 12]` where `onnx.reference.ReferenceEvaluator` on that same model returns `[5, 3, 0]`.

The correct mapping is `s![(norm_end + 1)..(norm_start + 1); step]` once both bounds are resolved against the dimension size, which also turns the `i64::MIN` sentinel into `lo = 0` for free.

## One genuine overflow, elsewhere in the same function

Auditing the rest of `calculate_dim_slice_output_len` as you suggested did turn up input-driven overflow, on `step` rather than on the index:

```rust
((range_len + step.abs() - 1) / step.abs())
```

- `step = i64::MIN` panics in `abs()` ("attempt to negate with overflow")
- `step = i64::MAX` panics on `range_len + step.abs()` ("attempt to add with overflow")

Both reproduce under `-C overflow-checks=on`. `steps` comes from the model unchecked, so these are reachable. Now computed as `(range_len as u64).div_ceil(step.unsigned_abs())`.

## Fix

Working on a branch now: correct ONNX-to-Burn reverse-step translation for both the tensor and Shape paths (including bounds resolved from the tensor's own dims when the axis size isn't static), the `unsigned_abs` change above, and integration tests covering the sentinel, a dynamic-size axis, and multi-step reverses.
